### ISSUE TITLE
Exception with libnetty_transport_native_epoll

### ISSUE THREAD

**Issue by merlimat** [netty/netty#7515]
### Actual behavior

I am experience some error in a Netty client running in Linux with native epoll transport type. 

It appears the client connection gets established but then there's an exception when writing on the socket. Possibly is associated with the fact the I'm passing a composite `ByteBuf` that will generate a `writev()` instead of regular `write()`?

~~I have upgraded to 4.1.18, but I'm now getting segfault in the epoll transport library (jvm log mentioned below). ~~ The segfault is solved in 4.1.19.

### Steps to reproduce

I'm reproducing it right now by using a CLI Pulsar client. From  the logs I can see the connections gets established, the initial handshake happens, but publishing the 1st message fails: 

```
2017-12-16 23:06:01,878 - INFO  - [pulsar-client-io-2-3:ConnectionPool@157] - [[id: 0xde74874f, L:/172.17.0.2:39044 - R:911b4f6a678e/172.17.0.2:6650]] Connected to server
2017-12-16 23:06:01,883 - INFO  - [pulsar-client-io-2-3:ProducerImpl@804] - [persistent://prop/cluster/ns/my-topic] [null] Creating producer on cnx [id: 0xde74874f, L:/172.17.0.2:39044 - R:911b4f6a678e/172.17.0.2:6650]
2017-12-16 23:06:02,119 - INFO  - [pulsar-client-io-2-3:ProducerImpl@825] - [persistent://prop/cluster/ns/my-topic] [standalone-2-0] Created producer on cnx [id: 0xde74874f, L:/172.17.0.2:39044 - R:911b4f6a678e/172.17.0.2:6650]
2017-12-16 23:06:02,250 - WARN  - [pulsar-client-io-2-3:ClientCnx@156] - [id: 0xde74874f, L:/172.17.0.2:39044 ! R:911b4f6a678e/172.17.0.2:6650] Exception caught: writev(..) failed: Bad address
io.netty.channel.unix.Errors$NativeIoException: writev(..) failed: Bad address
	at io.netty.channel.unix.Errors.newIOException(Errors.java:117)
	at io.netty.channel.unix.Errors.ioResult(Errors.java:138)
	at io.netty.channel.unix.FileDescriptor.writev(FileDescriptor.java:148)
	at io.netty.channel.epoll.AbstractEpollStreamChannel.writeBytesMultiple(AbstractEpollStreamChannel.java:323)
	at io.netty.channel.epoll.AbstractEpollStreamChannel.writeBytes(AbstractEpollStreamChannel.java:265)
	at io.netty.channel.epoll.AbstractEpollStreamChannel.doWriteSingle(AbstractEpollStreamChannel.java:472)
	at io.netty.channel.epoll.AbstractEpollStreamChannel.doWrite(AbstractEpollStreamChannel.java:457)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:856)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.flush0(AbstractEpollChannel.java:469)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush(AbstractChannel.java:823)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.flush(DefaultChannelPipeline.java:1296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:776)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:802)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:814)
	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:794)
	at org.apache.pulsar.client.impl.ProducerImpl$WriteInEventLoopCallback.run(ProducerImpl.java:452)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:311)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
```

From these logs, I can tell the client is able to connect to server, exchange initial handshake with protobuf (`Connect` -> `Connected`), then  creates producer (another exchange of protobuf objects). After that, we try to send 1st message, and that tries to write a composite ByteBuf that contains 2 actual buffers, for headers and actual payload.

### Minimal yet complete reproducer code (or URL to code)

I'll try to work on one.

### Netty version

Using `netty-all` artifact.

Netty 4.1.12 does not have this problem. 
Netty 4.1.13 to 4.1.17 : Exception
Netty 4.1.18 : Segfault. JVM log at https://gist.github.com/merlimat/e4ed1f649ff2dde924978c7d47e60f2c
Netty 4.1.19: No segfault but same exception

### JVM version (e.g. `java -version`)
`OpenJDK Runtime Environment (build 1.8.0_131-8u131-b11-2ubuntu1.16.04.3-b11)`

### OS version (e.g. `uname -a`)
Ubuntu 16.04.2
`Linux 911b4f6a678e 4.9.49-moby #1 SMP Fri Dec 8 13:40:02 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by merlimat** [comment]
Nevermind. Just saw there was a fix in af2f343648f4cb3deca5314174e0e579f9fec846 included for next release 4.1.19. I'll test with that.

**Comment by johnou** [comment]
@scottmitch

**Comment by normanmaurer** [comment]
Yep... will announce the release tomorrow. @merlimat all good now?

**Comment by merlimat** [comment]
@normanmaurer Just tested with 4.1.19 tag. The segfault is resolved but I'm now back to the exception I reported above for version 4.1.17: ` writev(..) failed: Bad address`.

I'll try to reproduce in a simple test.

**Comment by normanmaurer** [comment]
@merlimat thanks keep me posted

**Comment by Scottmitch** [comment]
isolated reproducer would be very helpful. Writing/flushing a single CompositeByteBuf with multiple ByteBuf components is a way to enter this code path.

> tries to write a composite ByteBuf that contains 2 actual buffers

What is the size of the 2 buffers?

**Comment by sschepens** [comment]
@merlimat any progress on this? we're affected when trying to use `pulsar-client` on netty greater than 4.1.12

**Comment by merlimat** [comment]
I found that it is related to the `DoubleByteBuf` class we're using. `DoubleByteBuf` is a read-only wrapper to hold 2 ByteBufs. It's less expensive that a `CompositeByteBuf` because it doesn't need to keep a list of buffer and also the object itself is recyclable.

The error only happens (starting from 4.1.13) when using direct buffers. I have a minimal client that reproduce the problem: https://gist.github.com/merlimat/799b8ba84b5987de369668a748aa0e32

I suspect the problem lies in the `DoubleByteBuf` implementation. There has to be some detail that I'm missing, I'll keep searching.

**Comment by sschepens** [comment]
@merlimat I think Pulsar's `DoubleByteBuf` is not playing well with native transports, I managed to fix this, but I don't know if it's the right fix, i created a pr https://github.com/apache/incubator-pulsar/pull/1056

**Comment by sschepens** [comment]
@normanmaurer @merlimat I found the breaking change in netty, though the change seems OK, it's bad behavior of `DoubleByteBuf` that's causing the issue.

`UnixChannelUtil` changed its `isBufferCopyNeededForWrite` logic from:
```java
!byteBuf.hasMemoryAddress() || !byteBuf.isDirect() || byteBuf.nioBufferCount() > iovMax
```
To:
```java
!byteBuf.hasMemoryAddress() && (!byteBuf.isDirect() || byteBuf.nioBufferCount() > iovMax)
```

Pulsar's `DoubleByteBuf` always returns `false` to `hasMemoryAddress()` but returns `true` to `isDirect` if it's components are direct, therefore, netty is not performing a copy and expects `DoubleByteBuf.nioBuffers` to return direct buffers, but `DoubleByteBuf` was always returning heap buffers.

I believe the changes I made in https://github.com/apache/incubator-pulsar/pull/1056 should be OK then.
